spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BinaryTreeLSTM.scala

```diff
@@ -0,0 +1,474 @@
+    (cp ++ lp, cg ++ lg)
+  }
+
+  override def updateParameters(learningRate: T): Unit = {
+    composer.updateParameters(learningRate)
+    leafModule.updateParameters(learningRate)
+  }
+
+  override def getParametersTable(): Table = {
+    val pt = T()
+    val t1 = composer.getParametersTable()
+    val t2 = leafModule.getParametersTable()
+    t1.keySet.foreach(key => pt(key) = t1(key))
+    t2.keySet.foreach(key => pt(key) = t2(key))
+    pt
+  }
+
+  override def zeroGradParameters(): Unit = {
+    composer.zeroGradParameters()
+    leafModule.zeroGradParameters()
+  }
+
+  override def reset(): Unit = {
+    composer.reset()
+    leafModule.reset()
+  }
+
+  override def hashCode(): Int = {
+    def getHashCode(a: Any): Int = if (a == null) 0 else a.hashCode()
+    val state = Seq(super.hashCode(), composer, leafModule)
+    state.map(getHashCode).foldLeft(0)((a, b) => 31 * a + b)
+  }
+
+  override def canEqual(other: Any): Boolean = other.isInstanceOf[BinaryTreeLSTM[T]]
+
+  override def equals(other: Any): Boolean = other match {
+    case that: BinaryTreeLSTM[T] =>
+      super.equals(that) &&
+        (that canEqual this) &&
+        composer == that.composer &&
+        leafModule == that.leafModule
+    case _ => false
+  }
+}
+
+object BinaryTreeLSTM {
+  def apply[@specialized(Float, Double) T: ClassTag](
+    inputSize: Int,
+    hiddenSize: Int,
+    gateOutput: Boolean = true,
+    withGraph: Boolean = true
+  )(implicit ev: TensorNumeric[T]): BinaryTreeLSTM[T] =
+    new BinaryTreeLSTM[T](inputSize, hiddenSize, gateOutput, withGraph)
+}
+
+class TensorTree[T: ClassTag](val content: Tensor[T])
+  (implicit ev: TensorNumeric[T]) extends Serializable {
+  def size: Array[Int] = content.size()
+
+  def children(index: Int): Array[Int] =
+    content.select(1, index).toBreezeVector().toArray.map(ev.toType[Int])
+
+  def addChild(parent: Int, child: T): Unit = {
+    breakable {
+      for (i <- 1 to size(1)) {
+        if (content(Array(parent, i)) == ev.zero) {
+          content.setValue(parent, i, child)
+          break()
+        }
+      }
+    }
+  }
+
+  def markAsRoot(index: Int): Unit = {
+    content.setValue(index, size(1), ev.negative(ev.one))
+  }
+
+  def getRoot: Int = {
+    for (i <- 1 to size(0)) {
+      if (ev.toType[Int](content(Array(i, size(1)))) == -1) {
+        return i
+      }
+    }
+
+    -1
+  }
+
+  def markAsLeaf(index: Int, leafIndex: Int): Unit = {
+    content.setValue(index, size(1), ev.fromType(leafIndex))
+  }
+
+  def leafIndex(index: Int): Int = {
+    ev.toType[Int](content(Array(index, size(1))))
+  }
+
+  def hasChild(index: Int): Boolean = {
+    ev.toType[Int](content(Array(index, 1))) > 0
+  }
+
+  def noChild(index: Int): Boolean = {
+    ev.toType[Int](content(Array(index, 1))) == 0
+  }
+
+  def exists(index: Int): Boolean = {
+    index >= 1 && index <= size(0)
+  }
+
+  def isPadding(index: Int): Boolean = {
+    ev.toType[Int](content(Array(index, 1))) == -1
```

Comment:
No, it should return false. Maybe we should add comment to restrict the padding value should not be 0.